File: hoffman-agents-node/examples/03_weather_benchmark/weather_benchmark.js
/**
 * Weather Benchmark — Markov World vs Classical Baselines
 *
 * Build a Markov world from weather CSV data and compare prediction
 * accuracy against persistence, climatology, and AR(1).
 *
 * Usage:
 *   node weather_benchmark.js [path/to/weather_berlin.csv]
 */
const fs = require('fs');
const path = require('path');
const { WorldBuilder } = require('../../src/index');

function loadCSV(filePath) {
  const text = fs.readFileSync(filePath, 'utf8').trim();
  const lines = text.split('\n');
  const headers = lines[0].split(',').map(h => h.trim());
  const rows = [];
  for (let i = 1; i < lines.length; i++) {
    const vals = lines[i].split(',').map(v => v.trim());
    if (vals.length === headers.length) {
      const row = {};
      headers.forEach((h, j) => { row[h] = vals[j]; });
      rows.push(row);
    }
  }
  return rows;
}

function floatOrNull(v) { return v === '' || v === 'None' ? null : parseFloat(v); }

function computeTempDirection(temp, horizon = 3, threshold = 0.5) {
  return temp.map((t, i) => {
    const target = i + horizon;
    if (target >= temp.length || t === null || temp[target] === null) return null;
    const change = temp[target] - t;
    if (Math.abs(change) <= threshold) return 'same';
    return change > 0 ? 'up' : 'down';
  });
}

function run(csvPath) {
  console.log('='.repeat(60));
  console.log('Weather Benchmark: Markov World vs Classical Baselines');
  console.log('='.repeat(60));

  if (!fs.existsSync(csvPath)) {
    console.error(`File not found: ${csvPath}`);
    console.error('Download from: https://github.com/anomalyco/conscious-agents/blob/main/data/weather_berlin.csv');
    process.exit(1);
  }

  const rows = loadCSV(csvPath);
  console.log(`\nLoaded ${rows.length} weather records`);

  const temp = rows.map(r => floatOrNull(r.temperature_2m));
  const pressure = rows.map(r => floatOrNull(r.surface_pressure));
  const humidity = rows.map(r => floatOrNull(r.relative_humidity_2m));
  const wind = rows.map(r => floatOrNull(r.wind_speed_10m));

  // Build data array (skip rows with nulls)
  const valid = [];
  for (let i = 0; i < rows.length; i++) {
    if (temp[i] !== null && pressure[i] !== null && humidity[i] !== null && wind[i] !== null) {
      valid.push([temp[i], pressure[i], humidity[i], wind[i]]);
    }
  }
  const n = valid.length;

  if (n < 100) {
    console.error('Too few valid rows after filtering');
    process.exit(1);
  }

  console.log(`Building Markov world...`);
  const world = new WorldBuilder()
    .addFeature('temp', 'minmax', 4)
    .addFeature('pressure', 'minmax', 4)
    .addFeature('humidity', 'minmax', 4)
    .addFeature('wind', 'minmax', 3)
    .build(valid);
  console.log(`  ${world.nStates} states`);

  const stateIds = world.stateIds;
  const dirs = computeTempDirection(temp);

  // Filter valid indices
  const validIdx = [];
  for (let i = 0; i < n; i++) {
    if (dirs[i] !== null) validIdx.push(i);
  }

  const split = Math.floor(validIdx.length * 0.8);
  const trainIdx = validIdx.slice(0, split);
  const testIdx = validIdx.slice(split);

  // Build direction map from training data
  const dirMap = {};
  for (const i of trainIdx) {
    const sid = stateIds[i], d = dirs[i];
    if (d === null) continue;
    if (!dirMap[sid]) dirMap[sid] = { up: 0, down: 0, same: 0 };
    dirMap[sid][d]++;
  }

  function predictDir(sid) {
    const counts = dirMap[sid];
    if (!counts) return null;
    let best = 'up', bc = 0;
    for (const [d, c] of Object.entries(counts)) { if (c > bc) { best = d; bc = c; } }
    return best;
  }

  // Markov evaluation
  let mCor = 0, mTot = 0;
  for (const i of testIdx) {
    const pred = predictDir(stateIds[i]);
    if (pred) { mTot++; if (pred === dirs[i]) mCor++; }
  }
  const markovAcc = mTot > 0 ? mCor / mTot : 0;

  // Majority class
  const trainCounts = { up: 0, down: 0, same: 0 };
  for (const i of trainIdx) { if (dirs[i]) trainCounts[dirs[i]]++; }
  const majority = Object.entries(trainCounts).sort((a, b) => b[1] - a[1])[0][0];
  const majorityAcc = trainCounts[majority] / Object.values(trainCounts).reduce((a, b) => a + b, 0);

  // Persistence
  let pCor = 0, pTot = 0;
  for (const i of testIdx) {
    const past = i - 6;
    if (past < 0) continue;
    const change = temp[i] - temp[past];
    const pred = Math.abs(change) <= 0.5 ? 'same' : change > 0 ? 'up' : 'down';
    pTot++; if (pred === dirs[i]) pCor++;
  }
  const persAcc = pTot > 0 ? pCor / pTot : 0;

  // AR(1)
  const arPairs = [];
  for (const i of trainIdx) {
    if (temp[i] !== null && temp[i + 3] !== null) arPairs.push([temp[i], temp[i + 3]]);
  }
  let arAcc = 0;
  if (arPairs.length > 10) {
    const xs = arPairs.map(p => p[0]), ys = arPairs.map(p => p[1]);
    const mx = xs.reduce((a, b) => a + b, 0) / xs.length;
    const my = ys.reduce((a, b) => a + b, 0) / ys.length;
    let num = 0, den = 0;
    for (let k = 0; k < xs.length; k++) { num += (xs[k] - mx) * (ys[k] - my); den += (xs[k] - mx) ** 2; }
    const a = den > 1e-12 ? num / den : 0;
    const b = my - a * mx;
    let arCor = 0, arTot = 0;
    for (const i of testIdx) {
      const predicted = a * temp[i] + b;
      const change = predicted - temp[i];
      const pred = Math.abs(change) <= 0.5 ? 'same' : change > 0 ? 'up' : 'down';
      arTot++; if (pred === dirs[i]) arCor++;
    }
    arAcc = arTot > 0 ? arCor / arTot : 0;
  }

  console.log(`\nBaselines:`);
  console.log(`  Majority class ('${majority}'):       ${(majorityAcc*100).toFixed(1)}%`);
  console.log(`  Persistence (6h):            ${(persAcc*100).toFixed(1)}%`);
  console.log(`  AR(1):                       ${(arAcc*100).toFixed(1)}%`);
  console.log(`  Markov World:                ${(markovAcc*100).toFixed(1)}%`);
  console.log(`  Improvement over majority:   ${(markovAcc/majorityAcc*100-100).toFixed(1)}%`);

  if (markovAcc > Math.max(majorityAcc, persAcc, arAcc)) {
    console.log(`\n  ✓ Markov world beats classical baselines`);
  }
}

const csvPath = process.argv[2] || path.join(__dirname, 'weather_berlin.csv');
run(csvPath);

File: hoffman-agents-python/examples/03_weather_benchmark/weather_benchmark.py
"""
Weather World Benchmark — Markov World vs Classical Baselines

Build a Markov world from weather data (temperature, pressure, humidity)
and compare conscious agent prediction against classical methods:
  - Persistence: the last N hours repeat
  - Climatology: most common outcome for this hour
  - AR(1): linear autoregression

Expected result:
  Markov World prediction accuracy >> all classical baselines
"""
from conscious_agent import WorldBuilder, World as CAWorld
import numpy as np
import csv
import math
import statistics
from pathlib import Path


def load_csv_rows(path):
    with open(path, newline="") as f:
        return list(csv.DictReader(f))


def compute_features(rows):
    temp = [float(r["temperature_2m"]) for r in rows]
    pressure = [float(r["surface_pressure"]) for r in rows]
    humidity = [float(r["relative_humidity_2m"]) for r in rows]
    wind = [float(r["wind_speed_10m"]) for r in rows]
    return temp, pressure, humidity, wind


def compute_temp_direction(temp, horizon=3, threshold=0.5):
    dirs = []
    for i in range(len(temp)):
        target = i + horizon
        if target >= len(temp):
            dirs.append(None)
        else:
            change = temp[target] - temp[i]
            if abs(change) <= threshold:
                dirs.append("same")
            elif change > 0:
                dirs.append("up")
            else:
                dirs.append("down")
    return dirs


def run_benchmark(csv_path=None):
    if csv_path is None:
        csv_path = Path(__file__).parent / "weather_berlin.csv"
    np.random.seed(42)
    print("=" * 60)
    print("Weather Benchmark: Markov World vs Classical Baselines")
    print("=" * 60)

    rows = load_csv_rows(csv_path)
    temp, pressure, humidity, wind = compute_features(rows)
    n = len(rows)
    print(f"\nLoaded {n} weather records")

    # Build features array for WorldBuilder
    data = np.column_stack([
        np.array(temp, dtype=float),
        np.array(pressure, dtype=float),
        np.array(humidity, dtype=float),
        np.array(wind, dtype=float),
    ])

    # Build Markov world
    print("Building Markov world...")
    world = (WorldBuilder()
        .add_feature("temp", "minmax", 4)
        .add_feature("pressure", "minmax", 4)
        .add_feature("humidity", "minmax", 4)
        .add_feature("wind", "minmax", 3)
        .build(data))
    print(f"  {world.n_states} states")

    # Get discretized state IDs from the world's state_ids
    state_ids = world.state_ids

    # Compute temperature direction labels
    dirs = compute_temp_direction(temp, horizon=3, threshold=0.5)

    # Build direction map from state → {up, down, same} counts (training only)
    split = int(n * 0.8)
    train_states = state_ids[:split]
    test_states = state_ids[split:]
    train_dirs = dirs[:split]
    test_dirs = dirs[split:]

    dir_map = {}
    for sid, d in zip(train_states, train_dirs):
        if d is None:
            continue
        if sid not in dir_map:
            dir_map[sid] = {"up": 0, "down": 0, "same": 0}
        dir_map[sid][d] += 1

    def predict_dir(sid):
        counts = dir_map.get(sid)
        if not counts:
            return None
        return max(counts, key=counts.get)

    # Evaluate Markov on test
    correct = total = 0
    for sid, actual in zip(test_states, test_dirs):
        if actual is None:
            continue
        pred = predict_dir(sid)
        if pred:
            total += 1
            if pred == actual:
                correct += 1
    markov_acc = correct / total if total > 0 else 0

    # Majority class baseline
    train_counts = {"up": 0, "down": 0, "same": 0}
    for d in train_dirs:
        if d:
            train_counts[d] += 1
    majority = max(train_counts, key=train_counts.get)
    majority_acc = train_counts[majority] / sum(train_counts.values())

    # Persistence baseline
    pers_correct = pers_total = 0
    for i in range(split, n):
        past = i - 6
        if past < 0 or dirs[i] is None or temp[past] is None:
            continue
        past_change = temp[i] - temp[past]
        if abs(past_change) <= 0.5:
            pred = "same"
        elif past_change > 0:
            pred = "up"
        else:
            pred = "down"
        pers_total += 1
        if pred == dirs[i]:
            pers_correct += 1
    pers_acc = pers_correct / pers_total if pers_total > 0 else 0

    # AR(1) baseline
    train_temp = temp[:split]
    pairs = [(train_temp[t], train_temp[t + 3])
             for t in range(len(train_temp) - 3)
             if train_temp[t] is not None and train_temp[t + 3] is not None]
    if len(pairs) > 10:
        xs = [p[0] for p in pairs]
        ys = [p[1] for p in pairs]
        mx = statistics.mean(xs)
        my = statistics.mean(ys)
        denom = sum((x - mx) ** 2 for x in xs)
        a = sum((x - mx) * (y - my) for x, y in zip(xs, ys)) / denom if denom > 1e-12 else 0
        b = my - a * mx

        ar_correct = ar_total = 0
        for i in range(split, n):
            if temp[i] is None or dirs[i] is None:
                continue
            predicted = a * temp[i] + b
            change = predicted - temp[i]
            if abs(change) <= 0.5:
                pred = "same"
            elif change > 0:
                pred = "up"
            else:
                pred = "down"
            ar_total += 1
            if pred == dirs[i]:
                ar_correct += 1
        ar_acc = ar_correct / ar_total if ar_total > 0 else 0
    else:
        ar_acc = 0

    # Print results
    print(f"\nBaselines:")
    print(f"  Majority class ('{majority}'): {majority_acc * 100:.1f}%")
    print(f"  Persistence (6h):           {pers_acc * 100:.1f}%")
    print(f"  AR(1):                      {ar_acc * 100:.1f}%")
    print(f"")
    print(f"  Markov World:               {markov_acc * 100:.1f}%")
    print(f"  Improvement over majority:  {(markov_acc / majority_acc - 1) * 100:+.1f}%")
    print(f"  Improvement over AR(1):     {(markov_acc / ar_acc - 1) * 100:+.1f}%")

    if markov_acc > max(majority_acc, pers_acc, ar_acc):
        print(f"\n  ✓ RESULT: Markov world beats all classical baselines")
    else:
        print(f"\n  ~ RESULT: Classical baselines competitive")

    return {
        "markov": markov_acc,
        "majority": majority_acc,
        "persistence": pers_acc,
        "ar1": ar_acc,
    }


if __name__ == "__main__":
    import sys
    csv_path = sys.argv[1] if len(sys.argv) > 1 else None
    run_benchmark(csv_path)
